README


                        JSPWiki documentation
                        =====================

                         $Revision: 1.28.2.4 $
                             $Date: 2006-11-09 18:49:15 $
                           $Author: jalkanen $

What is it?
-----------

JSPWiki is a simple (well, not any more) WikiWiki clone, written in Java
and JSP.  A WikiWiki is a web site which allows anyone to participate
in its development.  JSPWiki supports all the traditional wiki features,
as well as very detailed access control and security integration using JAAS. 


    Copyright (C) 2001-2006 JSPWiki development team

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public License
    as published by the Free Software Foundation; either version 2.1
    of the License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this program; if not, write to the Free
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA


The license file can be found in doc/LICENSE.

(No, the LGPL does not mean that all the content in your new wiki will
be under the LGPL, anymore than any text you write using Emacs becomes
GPL.)

One small request I have.  If you really, really like this software,
donate a small amount of money to the World Wildlife Foundation
(WWF), Free Software Foundation (FSF), or Electronic Freedom
Foundation (EFF), or whichever charity is close to your heart.

The world doesn't need to become a bad place.

Oh yeah, and if you run a Wiki based on this software, let me know.
I'm always interested in hearing from my users =).


Pre-requirements
----------------

Okay, so you wanna Wiki?  You'll need the following things:

REQUIRED:

* A JSP engine that supports Servlet API 2.3.  I recommend Tomcat from
  http://jakarta.apache.org/tomcat for a really easy installation.
  Tomcat 4.x or later is supported, however, we strongly recommend
  Tomcat 5.x or later.

* Some previous administration experience...  If you've ever installed
  Apache or any other web server, you should be pretty well off.

* And of course, a server to run the JSP engine on.

* JDK 1.4+.  JDK 1.5 is recommended.


OPTIONAL:

* JavaMail package from java.sun.com, if you want to use log4j mailing
  capabilities.  You'll also need the Java Activation Framework.

* RCS versioning system installed and available in your PATH if you
  want to use RCS versioning.  It is also possible (and on Windows, it
  is even recommended) to use the VersioningFileProvider, which does
  not require RCS.


Really simple installation
--------------------------

This section is meant for you, if you just want to have a really quick
install without much worries.  If you want to have a more complicated
installation (with more power as to how to do things), 
check out the "Install" section below.

Since JSPWiki 2.1.153, JSPWiki comes with a really simple installation
engine.  Just do the following:

1) Install Tomcat from http://jakarta.apache.org/ (or any other servlet
   container)

2) Drop the JSPWiki.war file in your <tomcat home>/webapps/ directory.
   Optional: Rename the JSPWiki.war if you want your URL to end with
   something else than /JSPWiki/.  For example, to get it called "wiki"
   just rename the "JSPWiki.war" file to "wiki.war".

3) Point your browser at http://<myhost>/JSPWiki/Install.jsp
   (Or, if you renamed it in the previous phase, use something like
   http://<myhost>/wiki/Install.jsp)

4) Answer a couple of simple questions

5) Restart your container

6) Point your browser to http://<myhost>/JSPWiki/ (or whatever name you
   chose in point 2)

That's it!


Advanced Installation
---------------------

First, install a JSP engine and a web server.  You can download Tomcat
from http://jakarta.apache.org/tomcat/.  This software was built
originally on top of Tomcat 5.5, but probably most other versions
work, too, and most other JSP containers.

After you've installed the engine and checked that it works, you just
make a subdirectory under $TOMCAT_HOME/webapps/.  For example, if you
want your application to be called 'MyWiki', just create a directory
called 'MyWiki', then extract all files from the war files into that
directory, preserving the directory structure.

Edit the WEB-INF/jspwiki.properties file to your liking.  The only
thing you need to set is the directory where you want your wiki pages
to be stored.  I recommend a directory that is regularly backed up.

Unzip the contents of jspwiki-corepages.zip into your newly created
directory.  You can find the rest of the documentation in the
JSPWiki-doc.zip file.

(Re)start tomcat.

Point your browser at http://<where your Tomcat is installed>/MyWiki/.
You should see the Main Wiki page.  See the next section if you want
to edit the pages =).


Installation with Windows
-------------------------

If you are using a version of Microsoft Windows, you'll find that
getting RCS up and running may be a bit problematic, and thus we
recommend the VersioningFileProvider.

All JSPWiki documentation is in the web; please see
http://doc.jspwiki.org/2.4/wiki/WindowsInstall

Upgrading from previous versions
--------------------------------

Please read ReleaseNotes.

All JSPWiki documentation is in the web; please see
http://doc.jspwiki.org/2.4/


Configuring JSPWiki
-------------------

All configuration information is kept in WEB_INF/jspwiki.properties.
The file is pretty well documented, so I recommend you read it through
first. 

There are also some auxiliary configuration files, such as filters.xml.

Also, there is a far more detailed documentation in the web:

http://doc.jspwiki.org/2.4/


Where to go next aka. where is the rest of the documentation?
-------------------------------------------------------------

The rest of the documentation is at

http://doc.jspwiki.org/2.4/


Acknowledgements
----------------

This product includes software developed by the Apache Software Foundation 
(http://www.apache.org/).  See doc/LICENCE.Apache for more information.

This product includes software developed by the OpenSymphony Group
(http://www.opensymphony.org/).  See doc/LICENSE.OpenSymphony for more
information. 

There have been many contributors so far.  Especially thanks go to
Andrew Jaquith, Dirk Fredericx, Murray Altheim, Olaf Kock, Mark
Rawling, Dan Frankowski, Terry Steichen, Chuck Smith, Christoph Sauer,
Erik Bunn, Niilo Neuvo, Alain Ravet, Kalle Kivimaa, Jeff Phillips,
Mahlen Morris, Paul Downes, Michael Gentry, John Volkar, Torsten
Hildebrandt, and Foster Schucker who have given me quite a lot of
support and/or code.  There are also a number of other people who have
generously given their time, and you can find hopefully all of their
names in the ChangeLog.


Contact
-------

JSPWiki authors can be contacted on the jspwiki-users mailing list.
You can see the instructions on http://www.jspwiki.org/wiki/JSPWikiMailingList.

The mailing list should also be your primary place to ask support questions.
There are many wise and knowledgeable people on that list.

If all else fails, feel free to send me (Janne Jalkanen, the lead developer) 
direct email, at jalkanen+jspwiki@ecyrd.com.


LICENSES
========

JSPWiki is licenced under the Lesser General Public License, version
2.1.  See LICENSE.

The various Jakarta libraries are licensed under the Apache Public
License.  See http://www.apache.org/LICENSE.txt

The search_highlight.js script is licensed under the X11 (MIT)
License:

COPYRIGHT AND PERMISSION NOTICE

Copyright (c) 2002  Stuart Langridge

All rights reserved.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, and/or sell copies of the Software, and to permit persons
to whom the Software is furnished to do so, provided that the above
copyright notice(s) and this permission notice appear in all copies of
the Software and that both the above copyright notice(s) and this
permission notice appear in supporting documentation.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT
OF THIRD PARTY RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
HOLDERS INCLUDED IN THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY
SPECIAL INDIRECT OR CONSEQUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER
RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

Except as contained in this notice, the name of a copyright holder
shall not be used in advertising or otherwise to promote the sale, use
or other dealings in this Software without prior written authorization
of the copyright holder.



ChangeLog
2010-08-29  Harry Metske <metskem@apache.org>

        * 2.8.5-svn-1
        
        * JSPWIKI-669 Support external properties in WikiEngine (idea from John McKinney)
          You can now override individual properties in jspwiki.properties using standard
          Java System properties.
        
2010-05-08  Harry Metske <metskem@apache.org>

        * 2.8.4-svn-12
        
        * JSPWIKI-392 and JSPWIKI-442 invalid asserted usernames
        
        * JSPWIKI-649 CachingProvider.deleteVersion leaves a stale m_historyCache
        
2010-04-23  Dirk Frederickx <brushed@apache.org>

        * 2.8.4-svn-11
        
        * JSPWIKI-627: Login and MyPrefs buttons not properly aligned
        in Chrome and Safari.

2010-04-08  Harry Metske <metskem@apache.org>

        * 2.8.4-svn-10
        
        * JSPWIKI-644: NPE after page externally modified, reported by Fabian Haupt.

2010-02-21  Janne Jalkanen <jalkanen@apache.org>

        * 2.8.4-svn-9

        * JSPWIKI-637: Fix for zh_CN translation, thanks to David Gao.

2010-01-31  Harry Metske <metskem@apache.org>

        * 2.8.4-svn-8
        
        * JSPWIKI-633: TextUtil.replaceEntities done multiple times, reported by Stefan Bohn.
        
2010-01-20  Janne Jalkanen <jalkanen@apache.org>

        * 2.8.4-svn-7
        
        * Improved exporter to also export from a FileSystemProvider repo without actually
        loading the WikiEngine.
        
2009-12-23 Harry Metske <metskem@apache.org>

        * 2.8.4-svn-6
        
         * JSPWIKI-621: check last fileseparator on config page, duplicate of JSPWIKI-424,
            reported by Rolf Schumacher. 

2009-12-15 Harry Metske <metskem@apache.org>

        * 2.8.4-svn-5
        
         * JSPWIKI-619: Include geronimo-web.xml in JSPWiki.war (reported by Jurgen Weber) 
         * JSPWIKI-618: Allow Administrators to override attachment.maxsize  (reported by Jurgen Weber) 

2009-12-06 Harry Metske <metskem@apache.org>

        * 2.8.4-svn-4
        
         * JSPWIKI-617: Attachments unsearchable after renaming a page 

2009-11-24  Janne Jalkanen <jalkanen@apache.org>

        * 2.8.4-svn-3
        
        * Fixed export to deal better with broken repositories: automatic renaming
        of case-conflicts (on systems where they are possible).
        
        * Export now properly exports in lower-case.
        
2009-11-23  Janne Jalkanen <jalkanen@apache.org>

        * 2.8.4-svn-2
        
        * Added export script (export.sh); disabled attachment export for now to test
        proper exporting.
        
2009-11-21 Harry Metske <metskem@apache.org>

        * 2.8.4-svn-1
        
         * JSPWIKI-612: fix UserManagement.jsp 's option in user select box 
         (I like to add the name of the person who discovered this bug, but
         the codepage ISO8859-1 doesn't allow that)
       
2009-09-26 Dirk Frederickx <brushed@apache.org>

        * 2.8.3-svn-19
        
         * JSPWIKI-591: # sign behind headlines is now hidden on the printed wiki page
       
        
2009-09-06 Harry Metske <metskem@apache.org>

        * 2.8.3-svn-18
        
        * JSPWIKI-597: Upgrade yuicompressor from 2.3.3 to 2.4.2 to solve 
        RuntimeExceptions when using OpenJDK.
        
                
2009-08-30 Harry Metske <metskem@apache.org>

        * 2.8.3-svn-17
                
        * Introduced the new PageViewPlugin written by Andre van Dalen.
        This plugin counts page views and presents the statistics on page views
        See http://www.jspwiki.org/wiki/PageViewPlugin for the complete
        documentation.
        
        * Minor javadoc error in WikiSecurityException
        
        * Commented out log4j logger stmt in jspwiki.properties.tmpl causing duplicate
        log entries during tests
        
2009-08-30 Dirk Frederickx <brushed@apache.org>

        * 2.8.3-svn-16
                
        * JSPWIKI-577: tabbed sections with many tabs with long names will float the 
        tabs to a next line. Fixed the 'jumping' behavior of clicking this wide tabs,
        which was caused by a superfluous bottom-border css style.

        * Open/Close of collapsible boxes is now only possible by clicking the +/- icon.
        Any clicks on the header-bar will not change the status of the 
        collapsible box anymore. E.G. in the Info tab, when looking at page 
        differences between 2 versions, you can now click the version drop-downs
        without changing the status of the collapsible box.

        * Implemented the standard css overflow treatment of the 'diff' body. Horizontal 
        scroll-bars will only be shown when looking at wide content.
        This is the same overflow treatment as the page content.

        
2009-08-25 Dirk Frederickx <brushed@apache.org>

        * 2.8.3-svn-15
        
        * JSPWIKI-580: Fixing the quirky handling of the textarea selection range
        in IEx browsers.

2009-08-23 Dirk Frederickx <brushed@apache.org>

        * 2.8.3-svn-14

        * JSPWIKI-480. Small improvement on Attachment sorting.
        Not using the attribute 'jspwiki:sortvalue' to allow sorting of 
        date values, both in attachment and info tabs.

        * JSPWIKI-574. Redirect from UserPreferences.jsp hangs when null.
        (which is the case eg after subsequent log in) 

        * JSPWIKI-590. Use relative paths to link the javascript routines.
        Avoid the use of getContextPath(), as the URL may have been rewritten
        external to jspwiki. (reported by Florian) 
        This fix only resolves the references to javascript routines.
        

2009-08-19 Harry Metske <metskem@apache.org>

        * 2.8.3-svn-13

        * Changed Exporter's dateformat to JSR-170's required ISO 8601 format 

2009-08-19 Dirk Frederickx <brushed@apache.org>

        * 2.8.3-svn-12

        * JSPWIKI-576. Styles of section [Edit] links fixed for IE.


src/com/ecyrd/jspwiki/tags/EditorTag.java
/* 
    JSPWiki - a JSP-based WikiWiki clone.

    Copyright (C) 2001-2004 Janne Jalkanen (Janne.Jalkanen@iki.fi)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2.1 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
package com.ecyrd.jspwiki.tags;

import java.io.IOException;

import javax.servlet.jsp.JspException;
import javax.servlet.jsp.PageContext;
import javax.servlet.jsp.tagext.BodyContent;

import com.ecyrd.jspwiki.WikiContext;
import com.ecyrd.jspwiki.WikiEngine;
import com.ecyrd.jspwiki.WikiPage;
import com.ecyrd.jspwiki.auth.authorize.DefaultGroupManager;

import org.apache.ecs.xhtml.*;
import org.apache.ecs.GenericElement;
import org.apache.ecs.ConcreteElement;

/**
 *  Creates an editor component with all the necessary parts
 *  to get it working.
 *  <p>
 *  In the future, this component should be expanded to provide
 *  a customized version of the editor according to user preferences.
 *
 *  @author Janne Jalkanen
 *  @since 2.2
 */
public class EditorTag
    extends WikiBodyTag
{
    private String m_submit  = "Save";
    private String m_preview = "Preview";
    private String m_cancel  = "Cancel";
    private String m_formName = "editForm";
    private String m_action = null;
    private boolean m_isGroupPage = false;
    
    public void setSubmit( String s )
    {
        m_submit = s;
    }

    public void setPreview( String s )
    {
        m_preview = s;
    }

    public void setCancel( String s )
    {
        m_cancel = s;
    }

    public void setName( String s )
    {
        m_formName = s;
    }
    
    public void setAction( String s )
    {
        m_action = s;
    }
    
    private GenericElement createInput( String name, String value )
    {
        input in = new input();
        in.setType("hidden");
        in.setName( name );
        in.setValue( value );

        return in;
    }

    private GenericElement createSubmit( String name, String value )
    {
        input in = new input();
        in.setType("submit");
        in.setName( name );
        in.setValue( value );

        return in;
    }

    private form createSimpleEditor()
    {
        WikiPage   page   = m_wikiContext.getPage();
        WikiEngine engine = m_wikiContext.getEngine();
        
        form f = new form();
        f.setName(m_formName);
 
        if( m_action != null ) 
        {
            f.setAction( m_action );
        }
        else if( m_wikiContext.getRequestContext().equals( WikiContext.COMMENT ) ||
                 "comment".equals(m_wikiContext.getHttpParameter("action")) )
        {
            f.setAction( m_wikiContext.getURL( WikiContext.COMMENT, page.getName() ));
        }
        else
        {
            f.setAction( m_wikiContext.getURL( WikiContext.EDIT, page.getName() ));
        }
        
        f.setMethod( "post" );
        f.setAcceptCharset( engine.getContentEncoding() );

        p para1 = new p();

        // Kludge to get preview working from Comment.jsp
        if( m_wikiContext.getHttpParameter("author") != null )
        {
            para1.addElement( createInput("author", m_wikiContext.getHttpParameter("author")) );
        }
        
        f.addElement(para1);

        para1.addElement( createInput("page", page.getName() ) );
        para1.addElement( createInput("action", "save" ) );
        para1.addElement( createInput("edittime",
                                      (String)pageContext.getAttribute("lastchange",
                                                                       PageContext.REQUEST_SCOPE ) ) );
       
        if( m_wikiContext.getRequestContext().equals("comment") )
        {
            para1.addElement( createInput("comment","true") );
        }
        
        return f;
    }

    private ConcreteElement getEditorArea()
    {
        textarea area = new textarea();

        area.setClass("editor");
        area.setWrap("virtual");
        area.setName("text");
        area.setRows( 25 );
        area.setCols( 80 );
        area.setStyle( "width:100%;" );
       
        if( m_wikiContext.getRequestContext().equals("edit") )
        {
            String content = m_wikiContext.getEngine().getText( m_wikiContext, m_wikiContext.getPage() );
            
            // This is a major HACK
            // If this is a new Group, inject a default ACL and member list
            if ( "".equals( content ) && m_isGroupPage )
            {
                String user = m_wikiContext.getCurrentUser().getName();
                String page = m_wikiContext.getPage().getName();
                String group = page.substring(DefaultGroupManager.GROUP_PREFIX.length());
                content = "[{SET members=" + user + "}]\n"
                        + "[{ALLOW edit " + group + "}]\n";
            }
            area.addElement( content );
        }
 
        return area;
    }
    /**
     *  Returns an edit button block.
     * 
     * @return
     */
    private ConcreteElement getButtons()
    {
        p para2 = new p();

        para2.addElement( createSubmit( "ok",      m_submit ) );
        para2.addElement( "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;" );
        para2.addElement( createSubmit( "preview", m_preview ) );
        para2.addElement( "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;" );
        para2.addElement( createSubmit( "cancel",  m_cancel ) );

        return para2;
    }
    
    public final int doWikiStartTag()
        throws IOException
    {
        // ARJ: this is a bit of a hack
        String pageName = m_wikiContext.getPage().getName();
        m_isGroupPage = pageName.startsWith( DefaultGroupManager.GROUP_PREFIX );
        
        return EVAL_BODY_TAG;
    }
       
    public int doEndTag() throws JspException
    {
        BodyContent bc = getBodyContent();
        
        form editor = createSimpleEditor();
    
        // 
        // If there is no body tag content, then we'll assume old
        // behaviour and append the stuff ourselves.
        //
        if( bc == null || bc.getString().length() == 0 )
        {
            editor.addElement( getEditorArea() );

            editor.addElement( getButtons() );            
        }
        else
        {
            editor.addElement( bc.getString() );
        }
        
        try
        {
            pageContext.getOut().print( editor.toString() );
        }
        catch( IOException e )
        {
            throw new JspException("Could not print Editor tag: "+e.getMessage() );
        }
        
        return EVAL_PAGE;
    }
}
